sklearn/model_selection/tests/test_split.py

```diff
@@ -430,19 +430,30 @@ def test_shuffle_kfold_stratifiedkfold_reproducibility():
         np.testing.assert_equal(list(cv.split(X, y)), list(cv.split(X, y)))
         np.testing.assert_equal(list(cv.split(X2, y2)), list(cv.split(X2, y2)))
 
+    # Tests to ensure consistent behaviour even when random_state is not set.
     kf = KFold(3, shuffle=True)
     skf = StratifiedKFold(3, shuffle=True)
-
-    for cv in (kf, skf):
+    kf1 = KFold(3, shuffle=True)
+    kf2 = KFold(3, shuffle=True)
+    skf1 = StratifiedKFold(3, shuffle=True)
+    skf2 = StratifiedKFold(3, shuffle=True)
+    for cvs in ((kf, kf1, kf2), (skf, skf1, skf2)):
         for data in zip((X, X2), (y, y2)):
+            # For the same initialization, splits should be same across
+            # multiple split calls, even when random_state is not set.
+            np.testing.assert_equal(list(cvs[0].split(*data)),
+                                    list(cvs[0].split(*data)))
+
+            # For different initialisations, splits should not be same when
+            # random_state is not set.
             try:
-                np.testing.assert_equal(list(cv.split(*data)),
-                                        list(cv.split(*data)))
+                np.testing.assert_equal(list(cvs[1].split(*data)),
```

Comment:
As @amueller pointed out I had a similar implementation of numpy.testing.assert_not_equal in https://github.com/scikit-learn/scikit-learn/pull/7946. I'd suggest we add that as a assert helper in sklearn/utils/testing.py.

@amueller I don't think your solution is future proof because you are in an edge case of array comparing when comparing arrays containing numpy arrays. Example:

```py
import numpy as np

arr = np.array([np.array([1, 2]), np.array([1, 2, 3])])
arr = arr 
# Returns: array([True, True], dtype=bool)
# with warning: DeprecationWarning: numpy equal will not check object identity in the future.
# The error trying to get the boolean value of the comparison result will be raised.
arr1 = np.array([np.array([1, 2]), np.array([1, 2, 3])])
# Returns: False
# with warning: DeprecationWarning: elementwise == comparison failed; this will raise an error in the future.
```
Basically the first case kind of works because it falls-back to comparing identity of the inner arrays when element comparison does not return a scalar. The second case returns False because inner arrays are not identical.
More details at https://github.com/numpy/numpy/issues/8323#issuecomment-263147507.